[LLVM][GPU] NVPTX specific passes for code generation

georgemitenkov opened the pull request:
This PR adds basic functionality to encode codegen (here,
for NVPTX) passes into optimization pipeline for GPUs. In
particular, this allows to infer address spaces and perform
other NVPTX-specific optimizations.

Corresponding tests were also added.

iomaganaris commented on test/unit/codegen/codegen_llvm_ir.cpp:
```diff
@@ -1648,4 +1648,47 @@ SCENARIO("GPU kernel body IR generation", "[visitor][llvm][gpu]") {
+
+            // Check target information.
+            // TODO: this may change when more platforms are supported.
+            std::regex data_layout(R"(target datalayout = \"e-i64:64-i128:128-v16:16-v32:32-n16:32:64\")");
```
Based on the above error I believe that this should be:
```suggestion
            std::regex data_layout(R"(target datalayout = \"e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64\")");
```

iomaganaris commented on src/codegen/llvm/llvm_utils.cpp:
```diff
@@ -61,6 +63,71 @@ static void run_optimisation_passes(llvm::Module& module,
+    // Hardcode target infromation for now. Change if necessary.
+    llvm::Triple triple("nvptx64-nvidia-cuda");
+    std::string subtarget = "sm_20";
+    std::string features = "+ptx60";
```
IIUC this selects the PTX ISA version. I guess that we can use something more recent like `ptx70`?
Not sure whether this makes any difference in the generated code and performance though. Maybe it's something to check.
Also from what I've seen selecting an invalid version gives the earliest PTX supported version.

iomaganaris reviewed (commented):
Overall LGTM! :+1:
I only suggested some small fixes and improvements.
With `magkanar/gpu-runner` and some extra fixes I managed to execute the test mod files used in the unit tests on the GPU

georgemitenkov commented on src/codegen/llvm/llvm_utils.cpp:
```diff
@@ -61,6 +63,71 @@ static void run_optimisation_passes(llvm::Module& module,
+
+    // Hardcode target infromation for now. Change if necessary.
+    llvm::Triple triple("nvptx64-nvidia-cuda");
+    std::string subtarget = "sm_20";
```
Agreed, I will adjust that!

georgemitenkov commented on src/codegen/llvm/llvm_utils.cpp:
```diff
@@ -61,6 +63,71 @@ static void run_optimisation_passes(llvm::Module& module,
+    std::unique_ptr<llvm::TargetMachine> tm;
+
+    // Hardcode target infromation for now. Change if necessary.
+    llvm::Triple triple("nvptx64-nvidia-cuda");
```
No, we can also target 32-bit: LLVM uses it as well. But  for our purposes the only difference would be  data layout and target triple

georgemitenkov commented:
@iomaganaris I addressed the comments, now we can select the target architecture and the subtarget in CLI, e.g. 
```
bin/nmodl mod/test.mod  --verbose debug -o mod llvm --ir --no-debug --opt-level-ir 3 gpu --name nvptx64 --target-chip sm_20
```

pramodk commented on src/codegen/llvm/target_platform.hpp:
```diff
@@ -32,11 +32,16 @@ class Platform {
+    /// Target chip for GPUs.
+    /// TODO: this should only be available to GPUs! If we refactor target
+    /// classes so that GPUPlatform <: Platform, it will be nicer!
+    const std::string subtarget_name = "sm_35";
```
@iomaganaris : considering we are now mostly running stuff on Volta GPUs, it's better to have this `sm_70`? (could be `sm_60` for Pascal generation).

iomaganaris replied:
Yes, I think `sm_70` makes sense since most HPC and local computer systems now use GPU with Compute Capability 7+

pramodk reviewed (approved):
LGTM

iomaganaris reviewed (commented):
LGTM
I just added a small suggestion and responded to Pramod's comment

iomaganaris commented on src/codegen/llvm/llvm_utils.cpp:
```diff
@@ -61,6 +63,85 @@ static void run_optimisation_passes(llvm::Module& module,
+
+    // Set data layout and target triple information for the module.
+    auto triple = triple_str.at(platform_name);
+    module.setDataLayout(data_layout_str.at(platform_name);
```
I missed closing a parenthesis here :innocent: 
```suggestion
    module.setDataLayout(data_layout_str.at(platform_name));
```

iomaganaris reviewed (approved):
LGTM :+1:
